### PR TITLE
feat: Add support for Directed Reads

### PR DESCRIPTION
**Pull request by asthamohta**
This PR adds code changes for Directed Reads feature.

### CODE REVIEW

File: google/cloud/spanner_v1/__init__.py [hunk 1]
```diff
@@ -79,6 +80,7 @@
 This value can only be used for timestamp columns that have set the option
 ``(allow_commit_timestamp=true)`` in the schema.
 """
+TransactionTypes = DirectedReadOptions.ReplicaSelection.Type
```
Comment (surbhigarg92) [inline google/cloud/spanner_v1/__init__.py:83]: Do we need to create this type ? Can't customers directly import `DirectedReadOptions.ReplicaSelection.Type` 

We are not using this anywhere in the source code

File: google/cloud/spanner_v1/snapshot.py [hunk 2]
```diff
@@ -224,12 +225,20 @@ def read(
...
+            or regions should be used for non-transactional reads or queries.
+
         :rtype: :class:`~google.cloud.spanner_v1.streamed.StreamedResultSet`
         :returns: a result set instance which can be used to consume rows.
 
         :raises ValueError:
             for reuse of single-use snapshots, or if a transaction ID is
             already pending for multiple-use snapshots.
+
+        :raises InvalidArgument: if directed_read_options is set for READ-WRITE Transaction or PDML.
```
Comment (surbhigarg92) [inline google/cloud/spanner_v1/snapshot.py:241]: This should be removed ?

File: google/cloud/spanner_v1/snapshot.py [hunk 3]
```diff
@@ -379,9 +395,17 @@ def execute_sql(
...
+            or :class:`dict`
+        :param directed_read_options: (Optional) Request level option used to set the directed_read_options
+            for all ReadRequests and ExecuteSqlRequests that indicates which replicas
+            or regions should be used for non-transactional reads or queries.
+
         :raises ValueError:
             for reuse of single-use snapshots, or if a transaction ID is
             already pending for multiple-use snapshots.
+
+        :raises InvalidArgument: if directed_read_options is set for READ-WRITE Transaction or PDML.
```
Comment (surbhigarg92) [inline google/cloud/spanner_v1/snapshot.py:408]: Here as well ?

File: google/cloud/spanner_v1/snapshot.py (hunk 2 above)
Comment (harshachinta) [reply google/cloud/spanner_v1/snapshot.py]: Done

File: google/cloud/spanner_v1/snapshot.py (hunk 3 above)
Comment (harshachinta) [reply google/cloud/spanner_v1/snapshot.py]: Done

File: google/cloud/spanner_v1/__init__.py (hunk 1 above)
Comment (harshachinta) [reply google/cloud/spanner_v1/__init__.py]: I think we don't need an additional type for this.
Instead customers can do the following,
```
from google.cloud.spanner_v1 import DirectedReadOptions
directed_read_options_for_request = {
    "include_replicas": {
        "replica_selections": [
            {
                "type_": 
    directed_read_options_for_request = {
        "include_replicas": {
            "replica_selections": [
                {
                    "type_": DirectedReadOptions.ReplicaSelection.Type.READ_ONLY,
                },
            ],
            "auto_failover_disabled": True,
        },
    },
            },
        ],
        "auto_failover_disabled": True,
    },
}
```

Does this look correct?

File: tests/unit/test_database.py [hunk 4]
```diff
@@ -2317,6 +2341,48 @@ def test_generate_read_batches_w_data_boost_enabled(self):
             "keyset": {"all": True},
             "index": self.INDEX,
             "data_boost_enabled": True,
+            "directed_read_options": None,
```
Comment (surbhigarg92) [inline tests/unit/test_database.py:2344]: We are adding directed read options in databoost test. If its intentional, rename the test .

File: tests/unit/test_database.py [hunk 5]
```diff
@@ -2534,6 +2603,43 @@ def test_generate_query_batches_w_data_boost_enabled(self):
             "sql": sql,
             "data_boost_enabled": True,
             "query_options": client._query_options,
+            "directed_read_options": None,
```
Comment (surbhigarg92) [inline tests/unit/test_database.py:2606]: same here

File: google/cloud/spanner_v1/__init__.py (hunk 1 above)
Comment (surbhigarg92) [reply google/cloud/spanner_v1/__init__.py]: yes

File: google/cloud/spanner_v1/types/spanner.py [hunk 6]
```diff
@@ -398,7 +398,7 @@ class DirectedReadOptions(proto.Message):
     .. _oneof: https://proto-plus-python.readthedocs.io/en/stable/fields.html#oneofs-mutually-exclusive-fields
 
     Attributes:
-        include_replicas (google.cloud.spanner_v1.types.DirectedReadOptions.IncludeReplicas):
+        include_replicas (google.cloud.spanner_v1.DirectedReadOptions.IncludeReplicas):
```
[the remark this answers is not in the document]
Comment (harshachinta) [reply google/cloud/spanner_v1/types/spanner.py]: Done. Thanks

**Review by surbhigarg92 — APPROVED** [review]

File: tests/unit/test_database.py (hunk 4 above)
Comment (harshachinta) [reply tests/unit/test_database.py:2344]: As discussed offline, the request assertion fails so we need to have this with its default value.

File: tests/unit/test_database.py (hunk 5 above)
Comment (harshachinta) [reply tests/unit/test_database.py:2606]: Same as above

**Pull request merged**